aiidalab_widgets_base/structures.py

```diff
@@ -1028,73 +1078,58 @@ def def_perpendicular_to_screen(self, _=None):
             )
             self.axis_p2.value = self.vec2str(versor.tolist())
 
-    def translate_dr(self, _=None):
+    @register_operator
+    def translate_dr(self, _=None, atoms=None, selection=None):
         """Translate by dr along the selected vector."""
-        atoms = self.structure.copy()
-
-        selection = self.selection
 
         atoms.positions[self.selection] += np.array(
             self.action_vector * self.displacement.value
         )
 
-        self.structure = atoms
-        self.selection = selection
+        return atoms, selection
```

Comment:
I am worried that the magic introduced with `@register_operator` might be obfuscating the actual behavior here too much. I understand having a guard against operating on empty sets, etc., but do we really have to also manipulate the state in the hidden wrapper function?

I think it would be much better to make explicit that the class state is changed here:
```suggestion
        self.atoms, self.selection = atoms, selection
```
If there is need for additional, checks, you could provide some utility function if needed:
```suggestion
        self.atoms, self.selection = self._update_atoms_selection(atoms, selection)
```

Alternatively, you could make the operator function truly stateless (either as `@staticmethod` or `@classmethod`) and then pass in the input, and return the output state.